### ISSUE TITLE
ModelCheckpoint does not work properly when monitored metric is only logged on rank 0

### ISSUE THREAD

**Issue by nicolasch96** [Lightning-AI/lightning#19072]
### Bug description

Hello,

I am opening this issue, following this other issue https://github.com/Lightning-AI/lightning/issues/15852. I noticed that ModelCheckpoint does not update 'best_model_score' of the checkpoint when the monitored metric is only logged on rank 0.

In other words, whenever we do self.log(keyMetric, keyMetricValue, sync_dist=True) we need to force 'rank_zero_only' to False in order to monitor the metric correctly. However, due to communication overhead, this can significantly slow down the training process, isn't it better to add a 'monitor_rank_zero_only' option to ModelCheckpoint?

It is not very straightforward and it took me some time to find the issue.

Thank you in advance

### What version are you seeing the problem on?

master

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 2.0):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>

**Comment by awaelchli** [comment]
Hi @nicolasch96 

> we need to force 'rank_zero_only' to False in order to monitor the metric correctly.

Do you have some evidence for this? `rank_zero_only` should not be set together with `sync_dist=True`. See [here](https://github.com/Lightning-AI/lightning/pull/19056/files).

**Comment by nicolasch96** [comment]
Sorry I might've described the issue in a wrong way. The problem is that if we log on ALL ranks, and use sync_dict=True

so something like this:

`keyMetricValue = self.logger.experiment.get_metric_history(self.logger.run_id, self.keyMetric)[-1].value if self.trainer.is_global_zero else 0
# Log key metric for early stopping
## mlflow logger is not passed to rank > 0, we cannot use get_metric_history on rank > 0.
self.log(self.keyMetric,
         keyMetricValue,
         on_step=False, 
         on_epoch=True, 
         logger=False, 
         sync_dist=True,
         reduce_fx='sum', 
         rank_zero_only=True,
        )`

**Comment by awaelchli** [comment]
> if we need yo to log on ALL ranks (for early stopping to work for example)

This works flawlessly without logging on every rank simultaneously. The early stopping callback is designed for this. To monitor a metric in the callback, you must use `self.log` or it won't be registered internally as a value to monitor.

```py
self.log(self.keyMetric,
     keyMetricValue,
     on_step=False, 
     on_epoch=True, 
     logger=False, 
     sync_dist=True,
     reduce_fx='sum', 
     rank_zero_only=True,  # <---------- DON'T DO THIS! Never do this
)
```
I can see from your code example that there is quite the misunderstanding here. 
Let's start from 0. 

You want to do this:
1. Train a model with multi-GPU
2. Log a metric
3. Monitor the logged metric for EarlyStopping
4. Use MLFlowLogger for tracking the logged metric

So here is how to do it.

```py
class Model(LightningModule):
		...

		def validation_step(self, batch, batch_idx): 
				...
				loss = loss_fn(...)
				self.log("val_loss", loss, sync_dist=True, reduce_fx='sum')


model = Model()
trainer = Trainer(
	devices=2,
	callbacks=EarlyStopping(monitor="val_loss"),
	logger=MLFlowLogger(...),
)
trainer.fit(model)
```

And that's it! Here I took the val loss as the example (reduce=sum may not make sense here), but you will replace it with whatever you have. No need to do funny business with `self.logger.experiment.get_metric_history`, no need to think about ranks or anything like that. Let me know what you think.

**Comment by nicolasch96** [comment]
Okay thank you very much! I guess it's clear. I will reopen if I find anything that doesn't work properly. I did notice

> self.log(self.keyMetric,
     keyMetricValue,
     on_step=False, 
     on_epoch=True, 
     logger=False, 
     sync_dist=True,
     reduce_fx='sum', 
     rank_zero_only=True,  # <---------- DON'T DO THIS! Never do this
)
I think all the problems come from this line, so I leave it to False. 

Thank you again
